lib/Dialect/TTNN/Utils/TransformUtils.cpp

```diff
@@ -31,4 +32,65 @@ Value getOrInsertDevice(PatternRewriter &rewriter, Operation *op) {
   rewriter.restoreInsertionPoint(currentInsertionPoint);
   return deviceOp.getResult();
 }
+
+// Helper method to insert a ToLayoutOp to convert the input operand to the
+// desired tensor layout, buffer type and memory layout.
+mlir::Value
+createToLayoutOp(Operation *op, Value inputValue, PatternRewriter &rewriter,
+                 Layout targetTensorLayout, BufferType targetTensorBufferType,
+                 std::optional<TensorMemoryLayout> targetTensorMemoryLayout) {
+  TTNNLayoutAttr inputLayoutAttr = getLayoutAttrFromValue(inputValue);
+
+  // Create element type based on tensor layout.
+  Type elementType = getElementType(rewriter.getContext(), targetTensorLayout,
+                                    inputLayoutAttr.getDataType());
+
+  // Create tensor memory layout attribute.
+  ttnn::TensorMemoryLayoutAttr outputMemLayoutAttr =
+      targetTensorMemoryLayout.has_value()
+          ? ttnn::TensorMemoryLayoutAttr::get(rewriter.getContext(),
+                                              targetTensorMemoryLayout.value())
+          : nullptr;
+
+  // Get the input operand type.
+  RankedTensorType inputToLayoutOpType =
+      mlir::cast<RankedTensorType>(inputValue.getType());
+
+  // Create the new encoding for the output tensor type.
+  TTNNLayoutAttr toLayoutOpResultEncoding =
+      inputLayoutAttr.withElementType(rewriter.getContext(), elementType)
+          .withBufferType(rewriter.getContext(), targetTensorBufferType)
+          .withMemoryLayout(rewriter.getContext(), outputMemLayoutAttr);
+
+  // Create the output result type with the new encoding.
+  RankedTensorType toLayoutOpResultType =
+      ttnn::utils::createRankedTensorTypeWithEncoding(inputToLayoutOpType,
+                                                      toLayoutOpResultEncoding);
+
+  // Create the output memory config attribute.
+  ttnn::MemoryConfigAttr outputMemConfigAttr = ttnn::MemoryConfigAttr::get(
+      rewriter.getContext(),
+      ttnn::BufferTypeAttr::get(rewriter.getContext(), targetTensorBufferType),
+      ttnn::ShardSpecAttr::get(
+          op->getContext(),
+          ttnn::ShapeAttr::get(rewriter.getContext(),
+                               toLayoutOpResultEncoding.getShardShape())),
+      outputMemLayoutAttr);
+
+  // Get the device value if the tensor output is not on the host.
+  Value deviceValue = targetTensorBufferType == ttnn::BufferType::SystemMemory
+                          ? nullptr
+                          : utils::getOrInsertDevice(rewriter, op);
```

Comment:
Tried this, but multiple tests started to fail once I changed the return type:
```
PLEASE submit a bug report to https://github.com/llvm/llvm-project/issues/ and include the crash backtrace.
Stack dump:
0.      Program arguments: ./build/bin/ttmlir-opt --ttir-to-ttnn-backend-pipeline --mlir-print-ir-after-all test/ttmlir/Dialect/TTNN/pooling/simple_maxpool2d.mlir
 #0 0x00007fb507bf895c llvm::sys::PrintStackTrace(llvm::raw_ostream&, int) (/opt/ttmlir-toolchain/lib/libLLVM.so.20.0git+0xe5495c)
 #1 0x00007fb507bf8d86 SignalHandler(int) Signals.cpp:0:0
 #2 0x00007fb506881520 (/lib/x86_64-linux-gnu/libc.so.6+0x42520)
 #3 0x0000560d57d3b7c1 mlir::Operation::getOpResultImpl(unsigned int) /opt/ttmlir-toolchain/include/mlir/IR/Operation.h:1006:5
 #4 0x0000560d57d3b7c1 mlir::Operation::getResult(unsigned int) /opt/ttmlir-toolchain/include/mlir/IR/Operation.h:402:54
 #5 0x0000560d57d3b7c1 mlir::OpTrait::OneTypedResult<mlir::tt::DeviceType>::Impl<mlir::tt::ttnn::GetDeviceOp>::getResult() /opt/ttmlir-toolchain/include/mlir/IR/OpDefinition.h:699:33
 #6 0x0000560d57d3b7c1 mlir::OpTrait::OneTypedResult<mlir::tt::DeviceType>::Impl<mlir::tt::ttnn::GetDeviceOp>::operator mlir::detail::TypedValue<mlir::tt::DeviceType>() /opt/ttmlir-toolchain/include/mlir/IR/OpDefinition.h:704:54
 #7 0x0000560d57d3b7c1 mlir::tt::ttnn::ToLayoutOp mlir::OpBuilder::create<mlir::tt::ttnn::ToLayoutOp, mlir::Type, mlir::Value, mlir::tt::ttnn::LayoutAttr&, mlir::tt::DataTypeAttr&, mlir::tt::ttnn::MemoryConfigAttr&, mlir::tt::ttnn::GetDeviceOp>(mlir::Location, mlir::Type&&, mlir::Value&&, mlir::tt::ttnn::LayoutAttr&, mlir::tt::DataTypeAttr&, mlir::tt::ttnn::MemoryConfigAttr&, mlir::tt::ttnn::GetDeviceOp&&) /opt/ttmlir-toolchain/include/mlir/IR/Builders.h:516:31
 #8 0x0000560d57d3b582 mlir::tt::ttnn::ToLayoutOp mlir::RewriterBase::replaceOpWithNewOp<mlir::tt::ttnn::ToLayoutOp, mlir::Type, mlir::Value, mlir::tt::ttnn::LayoutAttr&, mlir::tt::DataTypeAttr&, mlir::tt::ttnn::MemoryConfigAttr&, mlir::tt::ttnn::GetDeviceOp>(mlir::Operation*, mlir::Type&&, mlir::Value&&, mlir::tt::ttnn::LayoutAttr&, mlir::tt::DataTypeAttr&, mlir::tt::ttnn::MemoryConfigAttr&, mlir::tt::ttnn::GetDeviceOp&&) /opt/ttmlir-toolchain/include/mlir/IR/PatternMatch.h:537:18
 #9 0x0000560d57d3b582 (anonymous namespace)::ToLayoutOpConversionPattern::matchAndRewrite(mlir::tt::ttir::ToLayoutOp, mlir::tt::ttir::ToLayoutOpAdaptor, mlir::ConversionPatternRewriter&) const /localdev/sdjordjevic/src/tt-mlir/lib/Conversion/TTIRToTTNN/TTIRToTTNN.cpp:234:14
#10 0x0000560d57d3adae mlir::OpConversionPattern<mlir::tt::ttir::ToLayoutOp>::matchAndRewrite(mlir::Operation*, llvm::ArrayRef<mlir::Value>, mlir::ConversionPatternRewriter&) const /opt/ttmlir-toolchain/include/mlir/Transforms/DialectConversion.h:545:3
#11 0x0000560d57c3c52b mlir::ConversionPattern::matchAndRewrite(mlir::Operation*, mlir::PatternRewriter&) const (./build/bin/ttmlir-opt+0x189452b)
#12 0x0000560d57c6bb18 void llvm::function_ref<void ()>::callback_fn<mlir::PatternApplicator::matchAndRewrite(mlir::Operation*, mlir::PatternRewriter&, llvm::function_ref<bool (mlir::Pattern const&)>, llvm::function_ref<void (mlir::Pattern const&)>, llvm::function_ref<llvm::LogicalResult (mlir::Pattern const&)>)::$_0>(long) PatternApplicator.cpp:0:0
#13 0x0000560d57c6961a mlir::PatternApplicator::matchAndRewrite(mlir::Operation*, mlir::PatternRewriter&, llvm::function_ref<bool (mlir::Pattern const&)>, llvm::function_ref<void (mlir::Pattern const&)>, llvm::function_ref<llvm::LogicalResult (mlir::Pattern const&)>) (./build/bin/ttmlir-opt+0x18c161a)
#14 0x0000560d57c3cdfa (anonymous namespace)::OperationLegalizer::legalize(mlir::Operation*, mlir::ConversionPatternRewriter&) DialectConversion.cpp:0:0
#15 0x0000560d57c3c5b0 mlir::OperationConverter::convert(mlir::ConversionPatternRewriter&, mlir::Operation*) (./build/bin/ttmlir-opt+0x18945b0)
#16 0x0000560d57c3d244 mlir::OperationConverter::convertOperations(llvm::ArrayRef<mlir::Operation*>) (./build/bin/ttmlir-opt+0x1895244)
#17 0x0000560d57c408b9 mlir::applyFullConversion(llvm::ArrayRef<mlir::Operation*>, mlir::ConversionTarget const&, mlir::FrozenRewritePatternSet const&, mlir::ConversionConfig) (./build/bin/ttmlir-opt+0x18988b9)
#18 0x0000560d57c40925 mlir::applyFullConversion(mlir::Operation*, mlir::ConversionTarget const&, mlir::FrozenRewritePatternSet const&, mlir::ConversionConfig) (./build/bin/ttmlir-opt+0x1898925)
#19 0x0000560d57d2d10e (anonymous namespace)::ConvertTTIRToTTNNPass::runOnOperation() /localdev/sdjordjevic/src/tt-mlir/lib/Conversion/TTIRToTTNN/TTI
```

Reply:
I've chatted with @sdjordjevicTT, the cause of this error are places where device is an optional operand, so this function is used like this
```cpp
auto/Value device = condition ? getOrInsertDevice(...) : nullptr;
```
If the return type of `getOrInsertDevice` is `ttnn::GetDeviceOp` and since `ttnn::GetDeviceOp` is a `Value` and `nullptr` is valid `Value` value, compiler will infer the most specific type of `?:` and that will be `ttnn::GetDeviceOp` in this case. The problem is that `nullptr` is dynamically not a valid value for `ttnn::GetDeviceOp`, so it will crash during the runtime. The solution here is to hint the compiler to treat the return type of value as `Value`, with `Value(getOrInsertDevice(...))`, or use
```cpp
Value device = nullptr;
if (condition) {
    device = getOrInsertDevice(...);
```
instead of ternary operator. Of course other option is to leave it as-is, I just find it awkward that function that has a name `getOrInsertDevice` returns `Value`, it can lead to some very hacky solutions like
```cpp
device = cast<ttnn::GetDeviceOp>(getOrInsertDevice(...));
```
if we ever need to query something from `GetDeviceOp`.

Reply:
I agree, I changed the return type to GetDeviceOp and hinted ternary conditions to return Value.